src/array/cpu/cpu_id_hash_map.cc

```diff
@@ -0,0 +1,194 @@
+/**
+ *  Copyright (c) 2019 by Contributors
+ * @file array/cpu/cpu_id_hash_map.cc
+ * @brief Class about CPU id hash map
+ */
+
+#include "cpu_id_hash_map.h"
+
+#include <dgl/runtime/parallel_for.h>
+#include <dgl/runtime/device_api.h>
+
+#include <cmath>
+
+using namespace dgl::runtime;
+
+namespace dgl {
+namespace aten {
+
+template <typename IdType>
+IdType CpuIdHashMap<IdType>::CompareAndSwap(IdType* ptr,
+    IdType old_val, IdType new_val) {
+  #ifdef _MSC_VER
+  if (sizeof(IdType) == 4) {
+    return _InterlockedCompareExchange(reinterpret_cast<LONG*>(ptr),
+      new_val, old_val);
+  } else if (sizeof(IdType) == 8) {
+    return _InterlockedCompareExchange64(reinterpret_cast<LONGLONG*>(ptr),
+      new_val, old_val);
+  } else {
+    LOG(FATAL) << "ID can only be int32 or int64";
+  }
+  #elif __GNUC__   // _MSC_VER
+    return __sync_val_compare_and_swap(ptr, old_val, new_val);
+  #else  // _MSC_VER
+    #error "CompareAndSwap is not supported on this platform."
+  #endif  // _MSC_VER
+}
+
+template <typename IdType>
+CpuIdHashMap<IdType>::CpuIdHashMap() : _hmap(nullptr), _mask(0) {
+}
+
+template <typename IdType>
+size_t CpuIdHashMap<IdType>::Init(IdArray ids, IdArray unique_ids) {
+  const IdType* ids_data = ids.Ptr<IdType>();
+  const size_t num = ids->shape[0];
+
+  CHECK_GT(num, 0);
+
+  size_t capcacity = 1;
+  capcacity = capcacity << static_cast<size_t>(1 + std::log2(num * 3));
```

Comment:
`std::ceil(std::log2(num * 3))`?

Reply:
and merge line 50, 51 into single line? like `const size_t capacity = 0x1 << ...`?